### DISCUSSION TITLE
System.Data.SqlClient Not Supported

### DISCUSSION THREAD

**Discussion by wcunnien** [datalust/seq-tickets#1637] [Help] [answered]
Using the latest SEQ install version, 2022.1, I have written a very simple Seq App that writes to a SQL Server database. I have tried several different .NET versions (4.x Framework, Standard 2.1, and .NET 6.0). No matter what I do, after installing the Seq App in to Seq, the app errors with:

System.PlatformNotSupportedException: System.Data.SqlClient is not supported on this platform.

I used https://github.com/johnmiller86/seq-apps-sqlserverexporter/tree/master/Seq.Apps.SQLServerExporter as an example.

Should I downgrade to 2021.4?

Thanks.

**Comment by nblumhardt** [comment]
Hi Bill, thanks for your message.

The problem is due to System.Data.SqlClient requiring platform-specific native libraries; i.e. it needs different binaries deployed for Windows, Linux, and so-on. Seq's app system assumes for the most part that everything is portable, though you can target `net5.0-windows` if you're targeting Windows hosts only.

If `net5.0-windows` isn't an option, I'd first consider porting to Microsoft.Data.SqlClient, instead of System.Data.SqlClient. Their APIs are very similar so the code should be reasonably straightforward to port from one to the other, but Microsoft.Data.SqlClient is later and has better .NET Core support. `net6.0` will be the best framework to target in that case.

The Microsoft.Data.SqlClient NUPKG (just a zip file) has `runtimes/unix/` and `runtimes/win/` folders, with (I think) `netcoreapp3.1` folders under each of them supplying native libraries for each platform. If Microsoft.Data.SqlClient doesn't work as-is, adding `<Include ..` tags to your CSPROJ pulling in the correct native libraries should sort it out.

I didn't spot a build script in the repo you linked; there's a Build.ps1 and appveyor.yaml in https://github.com/datalust/seq-app-htmlemail which you might be able to adapt, if you need an example of how the build/pack process should work.

Let me know if this helps!

Nick

**Reply by billcunnien** [reply]
Thank you for the response!!

I switched to the newer client (Microsoft.Data.SqlClient) and using the seq-app-htmlemail as a model for the project file, I was able to generate a nuget package that had everything and the kitchen sink.

![image](https://user-images.githubusercontent.com/107622326/184648426-5b265706-ee1d-497d-a539-31234c80ed0e.png)

After uploading the package and updating the app in SEQ, I **_still_** get the following error:

**_System.PlatformNotSupportedException: Microsoft.Data.SqlClient is not supported on this platform._**

This super simple database writing app has become extremely difficult to run properly in SEQ. I have 17 unit tests around the functionality (a bit of overkill, IMO) and all pass, so I am quite confident that the app runs properly; however, it will not run in SEQ.

Here are the guts of the project file:

<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFramework>net6.0</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
		<GeneratePackageOnBuild>True</GeneratePackageOnBuild>
		<Version>1.0.23</Version>
		<Title>Credo Event Archive</Title>
		<Description>SeqApp that will archive events to a database table.</Description>
		<PackageIcon>database.png</PackageIcon>
		<PackageTags>seq-app</PackageTags>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Microsoft.Data.SqlClient" Version="5.0.0" />
		<PackageReference Include="Seq.Apps" Version="2021.4.0" />
	</ItemGroup>

	<ItemGroup>
		<None Include="./database.png" Pack="true" Visible="false" PackagePath="" />
		<None Include="./bin/$(Configuration)/$(TargetFramework)/Publish/**/*" Exclude="./bin/$(Configuration)/$(TargetFramework)/Publish/Credo.SeqEventArchive.dll;./bin/$(Configuration)/$(TargetFramework)/Publish/Seq.Apps.dll;./bin/$(Configuration)/$(TargetFramework)/Publish/Serilog.dll" Pack="true" PackagePath="lib/$(TargetFramework)" />
	</ItemGroup>

</Project>

Anything else I can do here?

Thanks!

**Reply by nblumhardt** [reply]
Hi Bill,

Thanks for the update! The screenshot shows a "runtimes" folder in the package; this very likely has the needed files, but for Seq to find them, they need to be moved up alongside the other DLLs.

Would it be possible to send a copy of the package and/or C# project through to us via `support@`? We can try re-packing things to get them loaded into Seq, and suggest some CSPROJ file modifications so that you can generate a working package yourself.

It'll also be necessary for us to know what platform you're hosting Seq on (Windows or Docker/Linux).

Best regards,
Nick

**Reply by billcunnien** [reply]
Thanks, Nick!

The DLL in the publish folder was a tenth of the size of the one in the runtimes folder. So, I adjusted my project file to pull from the runtimes and not the publish folder for that file. I am now getting past that pesky "not supported" error. That gives me confidence that I have the app and resulting package setup properly.

However, if I run the app on an entry I am now getting the following error:

Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: HTTP Provider, error: 0 - )

Of course, I am getting our network folks involved since this seems like an environmental thing and not necessarily a SEQ issue. I am bringing it up in case you are aware of any obvious thing I may need to check.

Thanks, again!

**Reply by billcunnien** [reply]
btw, I restarted the SEQ service and re-executed the app and I received the following error:

Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: F. Path '', line 0, position 0.

This error was followed by a ton of logging around the SQL connection. Newtonsoft is not part of the app. It seems like it is trying but still something is blocking the way.

**Reply by billcunnien** [reply]
Any idea why I would be getting a fatal memory error? Thanks.

**Reply by nblumhardt** [reply]
Hi Bill,

We'll need quite a bit more info to diagnose this one - letting us know what platform you're targeting, along with all of the error information, and ideally a copy of the package, would help get us pointed in the right direction.

Best regards,
Nick

**Reply by billcunnien** [reply]
Supplied details in a support email. Thanks.

**Reply by nblumhardt** [reply]
Thanks for the follow-up, Bill. We haven't been able to track this one down (SPAM filter or otherwise); is it possible the attachment is triggering an email filter at your end? It might be necessary to upload your package to Dropbox and post a link; sorry about the hassle!

**Reply by billcunnien** [reply]
Tried a followup email with one zip file. Hopefully that makes it through. If not, I will come up with another option.

**Reply by nblumhardt** [reply]
Received, thank you 👍

**Reply by billcunnien** [reply]
Hey Nick, were you able to find out what I did wrong? Thanks!

**Reply by nblumhardt** [reply]
Hi Bill, sorry about the slow turnaround; I'm still working through it, I'll send an update shortly.

**Reply by nblumhardt** [reply]
Success! I needed to copy both the correct Microsoft.Data.SqlClient, and Microsoft.Data.SqlClient.SNI DLLs into the package (before doing that, I hit the same Access Violation exception you reported).

I'm now trying to get the CSPROJ to package this stuff up properly at build time, I'll send my results back to you by email.

**Reply by billcunnien** [reply]
Thanks for your help, Nick! That inclusion of the SNI assembly was key.

**Comment by billcunnien** [comment, marked as answer]
Include the SNI assembly in the package file

**Reply by nblumhardt** [reply]
Also relevant: https://github.com/datalust/seqcli/issues/252
